### ISSUE TITLE
Proposal: add the ability to extend with custom validation rules

### ISSUE THREAD

**Issue by jormaechea** [IBM/openapi-validator#243]
This issue is for check if you'd be interested in adding this feature.

**Important:** The rest of the issue will use the following case, where we have some (made up) custom logic where we can add links to documentations for each operation using the `x-docs` extension.
```yaml
paths:
  /pets:
    get:
      x-docs:
        - 'https://example.com/docs/pet'
        - 'https://example.com/docs/pet-owner'
```

## About the feature

The idea is to be able to add business-logic custom validations. For example, to validate a custom extension, to enforce some properties to be required or to check for case convention.

## Implementation approach

I think that the best option is to extend this with npm packages.
We can use a plugin system where people can code their own plugins and register them in the `.validaterc` file. Something like the following:

```json
# .validaterc
{
  // other configurations...,
  plugins: [
    'custom-package-for-validations',
    '@prefixed/npm-packages',
    ['some-package', 'myNamedExport']
  ]
}
```

## Implementation details

Each plugin should export a function (or named exports of type function) that will be called from the core (this package). For example:

```js
const { isPlainObject } = require('lodash');

const docsExtensionName = 'x-docs';

module.exports = (specification, messageCarrier, configuration) => {

	// Perform custom validation where, for example:
	if(!specification.paths)
		return;

	for(const [path, pathData] of Object.entries(specification.paths)) {
		for(const [operation, operationData] of Object.entries(pathData)) {

			if(operation.startsWith('x-') || !operationData[docsExtensionName])
				continue;

			if(!Array.isArray(operationData[docsExtensionName])) {
				messageCarrier.addMessage(
					['paths', path, operation, docsExtensionName].join('.'),
					'Docs extension must be an array if present',
					'error'
				);
				continue;
			}

			operationData[docsExtensionName].forEach((docItem, docIndex) => {
				if(typeof docItem !== 'string') {
					messageCarrier.addMessage(
						['paths', path, operation, docsExtensionName, docIndex].join('.'),
						'Docs extension items must be string',
						'error'
					);
				}
			});
		}
	}
};
```

Other details:
- Each plugin will be called in the same order they were registered
- Each plugin will be called whether or not a previous plugin reported an error
- Async plugins may be supported
- If a plugin throws or rejects, validation should be reported as error
- Report may or may not be separated for each plugin

---

What do you think?

**Comment by mkistler** [comment]
We have recently integrated the Spectral validator and exposed its native support for custom rules.

https://github.com/ibm/openapi-validator#custom-spectral-rules

Spectral custom rules for many common validations can be written in yaml without any extra code required.  But there is also the ability to invoke javascript code on specific points in the API document for more complex validation tasks.

Please check this out ... I think this will meet your needs.

I'm going to close this issue as "complete", but if you have specific needs that aren't addressed by the Spectral custom rules please feel free to open a new issue.

**Comment by jormaechea** [comment]
Haven't seen that. I'll take a look at spectral and get back to you if it's needed.

Thanks!

**Comment by jormaechea** [comment]
@mkistler I've been digging around spectral and I have a question:
What's the end game for this package? Is it gonna end up as a spectral ruleset? Because if you can achieve anything with spectral, I don't see a good reason to mantain all the validation code that's prone to bugs that spectral rules may natively avoid.

**Comment by mkistler** [comment]
@jormaechea I don't think we know yet where we're going to end up.  There is certainly some duplication in capability here so right now we're focused on reducing that.  But I think it is quite possible that we find some validations that aren't amenable to Spectral rules and thus we retain in our own validator code base.  We'll see.
